Avoid auto creation of indexes in concat

TomNicholas opened the pull request:
If we create a `Coordinates` object using the concatenated `result_indexes`, and pass that to the `Dataset` constructor, we can explicitly set the correct indexes from the start, instead of auto-creating the wrong ones and then trying to overwrite them with the correct indexes later (which is what the current implementation does).

- [x] Possible fix for #8871
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

TomNicholas commented:
All the tests in `test_concat.py` pass, but there is this change causes one failure in `test_groupby.py`:

```python
FAILED xarray/tests/test_groupby.py::test_groupby_drops_nans - ValueError: no coordinate variables found for these indexes: {'id'}
```

TomNicholas commented on xarray/core/concat.py:
```diff
@@ -668,6 +681,7 @@ def get_indexes(name):
         result_indexes[dim] = index
 
     # TODO: add indexes at Dataset creation (when it is supported)
+    # TODO: do we actually need this step anymore?
     result = result._overwrite_indexes(result_indexes)
```
Commenting this line out does not break any tests in `test_concat.py`, which implies that it isn't even being used now.

benbovy replied:
Oops replied in https://github.com/pydata/xarray/pull/8872#discussion_r1539106484.

benbovy commented on xarray/core/concat.py:
```diff
@@ -665,10 +678,6 @@ def get_indexes(name):
         else:
             index_vars = index.create_variables()
         result[dim] = index_vars[dim]
-        result_indexes[dim] = index
-
-    # TODO: add indexes at Dataset creation (when it is supported)
-    result = result._overwrite_indexes(result_indexes)
```
Removing those lines don't break any existing test because the line above `result[dim] = index_vars[dim]` actually re-creates a default PandasIndex when assigning the new `dim` variable. However, this unnecessarily re-creates a new index (or re-wrap an existing one) and this may not work in the future if we allow passing a custom xarray index as `dim` argument to `concat`.

It would be better to explicitly add *both* `index` and `index_vars` to `result`. Best way would be to assign them to `result_indexes` and `coord_vars` respectively before constructing the `Coordinates` object and then the `result` object, unless there are cases where `result.drop_vars(unlabeled_dims)` would delete the index coordinate.

TomNicholas replied:
I think I've addressed your comment now @benbovy

TomNicholas commented:
#8884 is now merged into this PR but all that should do is make the `MergeError` I'm currently seeing get emitted slightly earlier.

TomNicholas commented on xarray/tests/test_concat.py:
```diff
@@ -978,6 +979,34 @@ def test_concat_str_dtype(self, dtype, dim) -> None:
 
         assert np.issubdtype(actual.x2.dtype, dtype)
 
+    def test_concat_avoids_index_auto_creation(self) -> None:
```
Reminder to myself: The reason this test didn't catch the problem described in https://github.com/TomNicholas/VirtualiZarr/issues/18#issuecomment-2023955860 is because this test checks that concatenating datasets that start without indexes stay without indexes, whereas that problem is from concatenating datasets with indexes but having the coordinate variables be silently replaced by `IndexVariable` objects created from the index data.

TomNicholas commented:
The groupby test is still [failing](https://github.com/pydata/xarray/actions/runs/8473061958/job/23216566180?pr=8872#step:9:217)

```python
FAILED xarray/tests/test_groupby.py::test_groupby_drops_nans - ValueError: no coordinate variables found for these indexes: {'id'}
```

This is weird, because it seems groupby creates a call to concat (on [line 1861 of `groupby.py`](https://github.com/pydata/xarray/blob/2120808bbe45f3d4f0b6a01cd43bac4df4039092/xarray/core/groupby.py#L1861)) in which it attempts to concatenate along `dim='id'`, but `id` exists as an index and as a data variable on the inputs, but not as a coordinate variable. The `Coordinates` constructor I added inside `concat` then complains about this situation. I can't tell whether this is a bug in groupby's input to `concat` that I've exposed (i.e. it should have made `'id'` a coord), or an edge case for `concat` to handle.

TomNicholas commented on xarray/core/concat.py:
```diff
@@ -646,29 +647,44 @@ def get_indexes(name):
+        coord_vars[dim] = index_vars[dim]
         result_indexes[dim] = index
+        unlabeled_dims = unlabeled_dims - set([dim])
+    else:
+        if dim in result_data_vars:
+            coord_vars[dim] = result_data_vars[dim]
+            result_data_vars.pop(dim)
```
I was able to fix the groupby test failure by adding this piece of code, but I'm still not sure that I should have needed to add this. See https://github.com/pydata/xarray/pull/8872#issuecomment-2026088168.

@dcherian your input here might be helpful.

TomNicholas commented:
Wait that's weird - the CI tests pass (though one doctest fails), but locally one of the pytest tests fails for me :/

TomNicholas commented:
Okay I've found out what this bug with `id` is, and it's not groupby's fault. This test reproduces the bug on this branch.

```python
def test_concat_data_variable_without_index(self):
    ds = Dataset({'id': 1.0})
    result = concat([ds, ds], dim='id')  # raises in here
    expected = Dataset({'id': ('id', [1.0, 1.0])})
    assert_identical(result, expected)
```
```
ValueError: no coordinate variables found for these indexes: {'id'}
```

What's happening is that inside `_dataset_concat` there is a call to `ds.expand_dims('id')` for all the inputs. This sees that the `id` variable exists as a 0-D data variable, and promotes it to a 1D variable backed by an index (by calling `create_default_index_implicit`). 

In my case I don't want this index to be created, but I'm not sure what the default behaviour of `expand_dims` should be. Should I either:

1. Add an option to `ds.expand_dims` for opting out of auto-creating the index (i.e. following on from #8711)?
2. Allow `ds.expand_dims` to create this index but then immediately discard it on the grounds that the inputs to `concat` had no indexes?

Basically: in the case that the concat dim doesn't already exist on the inputs, should an index be created along that new dim by default?

---

I'm starting to think that we should add some state machine hypothesis tests like the ones @dcherian is adding in #8658, which specifically check that:

1) At no point do we have an entry in `._indexes` without a corresponding coordinate variable,
2) The set of indexes across some input datasets is never expanded silently by operations like `concat`/`expand_dims`.

benbovy commented:
> what the default behaviour of expand_dims should be

A third option may be to keep a consistent behavior xarray-wise and expose a global option to (temporarily) change it, e.g.,

```python
with xr.set_options(default_index_type=xr.NO_INDEX):
    # doesn't create any index along the new "x" dimension
    xr.concat(objs, "x")


# The `xr.NO_INDEX` sentinel value is used here instead of `None` to avoid
# confusion with the behavior of, e.g., `xr.Coordinates(coords={...}, indexes=None)`
# that creates default indexes for dimension coordinates.
```

TomNicholas commented:
> A third option may be to keep a consistent behavior xarray-wise and expose a global option to (temporarily) change it, e.g.,

This seems like a lot, but it might make the most sense. Whether or not to create a new index whenever a 1D coordinate is created during another operation is something that will come up during `concat` (and `combine_nested`), `expand_dims`, maybe `stack`, kind of in constructors too?

I'm happy to implement whatever you think makes most sense @benbovy , as I would like to be able to merge this.

TomNicholas commented:
@benbovy last meeting @shoyer and I had a conversation about this, and thought that perhaps a new function kwarg should be implemented first, as a stepping stone towards a global flag to change the default of that kwarg later.

TomNicholas commented:
This PR now builds on top of #8960. That seemed to fix the groupby error (I think through fixing a buggy edge case in concat rather than actually adding the `create_1d_index` kwarg).

I'm almost there, but for for one weird test where I check that an `UnexpectedDataAccess` error is raised when auto-creating a new 1D coordinate index from my `ConcatenatableArray` class. Somehow the index is being succesfully created, pulling out the numerical value that is supposed to be hidden inside `ConcatenatableArray`, without triggering the exception. This baffles me how the indexing code has managed to get that number out.

TomNicholas commented on xarray/core/concat.py:
```diff
@@ -646,29 +647,44 @@ def get_indexes(name):
-    result.encoding = result_encoding
 
-    result = result.drop_vars(unlabeled_dims, errors="ignore")
+    coord_vars = {
```
Done in https://github.com/pydata/xarray/pull/8872/commits/bde9f2bd4a0831166b2cf08604e689a5a2c9184d

TomNicholas commented:
Something weird is going on with the doctest failure - somehow my changes to `concat` have caused `.to_stacked_array()` to return a coordinate as a string dtype instead of an object dtype, but without failing any tests...

https://github.com/pydata/xarray/actions/runs/8852520437/job/24311394367?pr=8872#step:7:82

TomNicholas commented:
> somehow my changes to concat have caused .to_stacked_array() to return a coordinate as a string dtype instead of an object dtype, but without failing any tests...

@benbovy would you mind taking a look at this?

The difference in behaviour occurs when you concatenate objects with a MultiIndex that points to string-dtype variables.

e.g. printing the `stackable_vars` that get concatenated along `z` inside `.to_stacked_array` in the failing doctest example:

```python
[<xarray.DataArray 'a' (x: 2, z: 3)> Size: 48B
array([[0, 1, 2],
       [3, 4, 5]])
Coordinates:
  * z         (z) object 24B MultiIndex
  * variable  (z) <U1 12B 'a' 'a' 'a'
  * y         (z) <U1 12B 'u' 'v' 'w'
Dimensions without coordinates: x, 
<xarray.DataArray 'b' (x: 2, z: 1)> Size: 16B
array([[6],
       [7]])
Coordinates:
  * z         (z) object 8B MultiIndex
  * variable  (z) <U1 4B 'b'
  * y         (z) object 8B nan
Dimensions without coordinates: x]
```

On this branch the resulting concatenated `variable` has a string dtype, but on `main` it has an object dtype. To me (who rarely ever uses MultiIndexes) this seems totally reasonable, and the behaviour on `main` of coercing `variable` in the result to object dtype seems wrong.

I tried to make up a simpler example, but the [simplest example in the docs of a MultiIndex](https://docs.xarray.dev/en/stable/user-guide/data-structures.html#multiindex-coordinates) already starts with an object dtype to hold strings, which already seems weird to me

```python
midx = pd.MultiIndex.from_arrays(
    [["R", "R", "V", "V"], [0.1, 0.2, 0.7, 0.9]], names=("band", "wn")
)
mda = xr.DataArray(np.random.rand(4), coords={"spec": midx}, dims="spec")
mda
```
```
<xarray.DataArray (spec: 4)> Size: 32B
array([0.642, 0.275, 0.462, 0.871])
Coordinates:
  * spec     (spec) object 32B MultiIndex
  * band     (spec) object 32B 'R' 'R' 'V' 'V'
  * wn       (spec) float64 32B 0.1 0.2 0.7 0.9
```
i.e. why is `band` here an object dtype instead of string? 

I also don't know why any of this would be affected by the changes in this PR 😕

TomNicholas commented:
Thanks @benbovy

> having the string dtype instead of the object dtype in the example above seems totally reasonable to me too (even better, actually).

Okay so I've added a regression test and changed the docstring to the new expected dtype.

TomNicholas commented on xarray/tests/test_concat.py:
```diff
@@ -999,6 +1001,56 @@ def test_concat_str_dtype(self, dtype, dim) -> None:
+            for _ in range(2)
+        ]
+
+        # TODO this should raise an error, but it doesn't
+        # It's somehow cheating it's way around the UnexpectedDataAccess error that should be raised
+        with pytest.raises(UnexpectedDataAccess):
+            combined = concat(datasets, dim="x", create_index=True)
```
```suggestion
        with pytest.raises(UnexpectedDataAccess):
            concat(datasets, dim="x", create_index=True)
```

outdated

TomNicholas commented:
Okay I think I'm happy with this PR now, ready for a full review!

dcherian commented on xarray/core/concat.py:
```diff
@@ -646,29 +656,37 @@ def get_indexes(name):
         result_indexes[dim] = index
+        unlabeled_dims = unlabeled_dims - set([dim])
 
     # TODO: add indexes at Dataset creation (when it is supported)
```
```suggestion
```

We're using "official API" now so this TODO seems irrelevant

dcherian commented on xarray/core/concat.py:
```diff
@@ -646,29 +656,37 @@ def get_indexes(name):
+    result = type(datasets[0])(result_data_vars, coords=coords_obj, attrs=result_attrs)
+    result.encoding = result_encoding
+
+    result = result.drop_vars(unlabeled_dims, errors="ignore")
```
This smells a bit, especially with `errors="ignore"`. Remove these from `coord_vars` instead?

Why is it needed? At the very least we should add a cleanup TODO.

TomNicholas replied:
Took a while to find but I managed to fix this in https://github.com/pydata/xarray/pull/8872/commits/b0e36124c04255378ee3f7928fb5d0276f628f89 🙂 

Should be a performance improvement too because now we don't create new dimension coordinates just to drop them at the end 🤯

TomNicholas replied:
(Aside - This is weird imo)

```python
In [2]: import xarray as xr

In [3]: ds = xr.Dataset(data_vars={"a": ("x", [0])})

In [4]: 'x' in ds.coords
Out[4]: False

In [5]: ds.coords['x']
Out[5]: 
<xarray.DataArray 'x' (x: 1)> Size: 8B
array([0])
Dimensions without coordinates: x
```

TomNicholas commented:
@benbovy another friendly ping here - I would love to get this merged soon because my virtualizarr package currently can't run off xarray `main` without this PR! (see https://github.com/TomNicholas/VirtualiZarr/issues/14#issuecomment-2083487645)

dcherian commented on xarray/core/concat.py:
```diff
@@ -162,6 +166,8 @@ def concat(
 
         If a callable, it must expect a sequence of ``attrs`` dicts and a context object
         as its only parameters.
+    create_index : bool, default is True
+        Whether to create new PandasIndex objects for any new 1D coordinate variables.
```
Consider renaming and rewriting this description? Perhaps `promote_to_index`?

TomNicholas replied:
How about `create_index_for_new_dim`? Because it's only going to happen when a 0D variable gains a dimension that it did not already have.

dcherian replied:
```
create_index_for_new_dim : bool, default is True
        Whether to create a new ``PandasIndex`` object when the objects being concatenated contain scalar variables named ``dim``.
```

TomNicholas replied:
That's better!

keewis commented on xarray/core/concat.py:
```diff
@@ -162,6 +166,8 @@ def concat(
 
         If a callable, it must expect a sequence of ``attrs`` dicts and a context object
         as its only parameters.
+    create_index : bool, default is True
```
```suggestion
    create_index : bool, default: True
```

keewis commented on xarray/tests/__init__.py:
```diff
@@ -234,6 +236,129 @@ def __getitem__(self, key):
+
+    def __array__(self, dtype: np.typing.DTypeLike = None):
+        raise UnexpectedDataAccess("Tried accessing data")
+
+    def __getitem__(self, key) -> ConcatenatableArray:
+        """Some cases of concat require supporting expanding dims by dimensions of size 1"""
+        # see https://data-apis.org/array-api/2022.12/API_specification/indexing.html#multi-axis-indexing
+        arr = self._array
+        for axis, indexer_1d in enumerate(key):
+            if indexer_1d is None:
+                arr = np.expand_dims(arr, axis)
+            elif indexer_1d is Ellipsis:
+                pass
+            else:
+                raise UnexpectedDataAccess("Tried accessing data.")
+        return ConcatenatableArray(arr)
+
+    def __array_function__(self, func, types, args, kwargs) -> Any:
+        if func not in HANDLED_ARRAY_FUNCTIONS:
+            return NotImplemented
+
+        # Note: this allows subclasses that don't override
+        # __array_function__ to handle ManifestArray objects
+        if not all(issubclass(t, ConcatenatableArray) for t in types):
+            return NotImplemented
+
+        return HANDLED_ARRAY_FUNCTIONS[func](*args, **kwargs)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs) -> Any:
+        """We have to define this in order to convince xarray that this class is a duckarray, even though we will never support ufuncs."""
+        return NotImplemented
+
+    def astype(self, dtype: np.dtype, /, *, copy: bool = True) -> ConcatenatableArray:
+        """Needed because xarray will call this even when it's a no-op"""
+        if dtype != self.dtype:
+            raise NotImplementedError()
+        else:
+            return self
+
+
```
should we move this to a separate module (which we can import in `xarray.tests`). `__init__` is getting a bit big

TomNicholas replied:
Moved to a new `arrays.py` module in https://github.com/pydata/xarray/pull/8872/commits/6d825e5d97088037933c3242d136b9c41488fd07

dcherian reviewed (approved):
LGTM

keewis commented on xarray/core/concat.py:
```diff
@@ -162,6 +166,8 @@ def concat(
 
         If a callable, it must expect a sequence of ``attrs`` dicts and a context object
         as its only parameters.
+    create_index_for_new_dim : bool, default is True
```
```suggestion
    create_index_for_new_dim : bool, default: True
```

keewis reviewed (approved):
I can't spot anything important, either.

It looks like the parameter type spec nit has been undone, though.

keewis commented on xarray/core/dataarray.py:
```diff
@@ -2586,8 +2586,8 @@ def expand_dims(
             inserted to the start of the result array.
-        create_index : bool, default is True
-            Whether to create new PandasIndex objects for any new 1D coordinate variables.
+        create_index_for_new_dim : bool, default is True
```
```suggestion
        create_index_for_new_dim : bool, default: True
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -4541,8 +4541,8 @@ def expand_dims(
             inserted to the start of the result array.
-        create_index : bool, default is True
-            Whether to create new PandasIndex objects for any new 1D coordinate variables.
+        create_index_for_new_dim : bool, default is True
```
```suggestion
        create_index_for_new_dim : bool, default: True
```

keewis reviewed (commented):
looks fine to me, as well.

TomNicholas commented:
Thank you both! ❤️
